src/condor_collector.V6/collector.cpp

```diff
@@ -874,104 +1068,162 @@ int CollectorDaemon::receive_query_cedar_worker_thread(void *in_query_entry, Str
-			std::string stats_config;
-			cad->LookupString("STATISTICS_TO_PUBLISH",stats_config);
-			if (stats_config != "stored") {
-				dprintf(D_ALWAYS,"Updating collector stats using a chained ad and config=%s\n", stats_config.c_str());
-				stats_ad = new ClassAd();
-				if (!filter_private_attrs) {
-					stats_ad->CopyFrom(*curr_rec->m_pvtAd);
-				}
-				daemonCore->dc_stats.Publish(*stats_ad, stats_config.c_str());
-				daemonCore->monitor_data.ExportData(stats_ad, true);
-				collectorStats.publishGlobal(stats_ad, stats_config.c_str());
-				stats_ad->ChainToAd(curr_rec->m_publicAd);
-				ad_to_send = stats_ad; // send the stats ad instead of the self ad.
+	// Perform the query
+	CollectorDaemon::collect_op op;
+	bool sending = false;
+	int more = 1;
+
+	for (int ix = 0; ix < num_adtypes; ++ix) {
+		const AdTypes whichAds = (AdTypes) query_entry->adt[ix].whichAds;
+
+		op.__filter__ = query_entry->adt[ix].constraint;
+		op.__skip_absent__ = query_entry->adt[ix].skip_absent;
+		op.__mytype__ = (query_entry->adt[ix].match_mytype) ? query_entry->adt[ix].tag : nullptr;
+		op.__resultLimit__ = MIN(query_entry->limit, query_entry->adt[ix].limit) - op.__numAds__;
+		op.__results__ = &results;
+		results.clear();
+
+		CollectorHashTable * table = nullptr;
+		if (whichAds == GENERIC_AD) {
+			table = collector.getGenericHashTable(query_entry->adt[ix].tag);
+		} else if (whichAds != ANY_AD) {
+			table = collector.getHashTable(whichAds);
+		}
+		if (table) {
+			collector.walkHashTable (*table,
+				[&op](CollectorRecord*cr){
+					return op.query_scanFunc(cr);
+				});
+		} else if (ix==0 && whichAds == ANY_AD) {
+			std::vector<CollectorHashTable *> tables = collector.getAnyHashTables(op.__mytype__);
+			for (auto table : tables) {
+				collector.walkHashTable (*table,
+					[&op](CollectorRecord*cr){
+						return op.query_scanFunc(cr);
+					});
+				if (op.__numAds__ >= op.__resultLimit__)
+					break;
 			}
+			num_adtypes = 1; // don't allow Any as part of a multi-table scan.
+		} else {
+			dprintf (D_ALWAYS, "Error no collector table for %s\n", query_entry->adt[ix].tag);
+			continue;
 		}
 
-		if (evaluate_projection) {
-			proj.clear();
-			projection.clear();
-			if (EvalString(ATTR_PROJECTION, cad, curr_rec->m_publicAd, projection) && ! projection.empty()) {
+		query_time += runtime.tick(tick_time);
+
+		if (results.empty())
+			continue;
+
+		// do first time initialization for sending ads
+		if ( ! sending) {
+			// send the results via cedar
+			sock->timeout(QueryTimeout); // set up a network timeout of a longer duration
+			sock->encode();
+			sending = true;
+		}
+		if (whichAds == STARTD_PVT_AD) { filter_private_attrs = false; }
+
+		bool evaluate_projection = proj_is_expr;
+		classad::References * whitelist = proj.empty() ? nullptr : &proj;
+		if (query_entry->is_multi) {
+			std::string tagattr(query_entry->adt[ix].tag); tagattr += ATTR_PROJECTION;
+			if (query->LookupString(tagattr, projection)) {
```

Comment:
If a type-specific projection attribute needs collector-side evaluation, it's silently ignored.

Reply:
yes, type specific projections can only be strings in the current code.